### ISSUE TITLE
Cannot profile code anymore

### ISSUE THREAD

**Issue by lizmat** [rakudo/rakudo#3824]
## The Problem

No profile files are written anymore.

## Expected Behavior

Profile files should be written.

## Actual Behavior

    $ raku --profile -e 'say "foo"'
    foo
    Writing profiler output to profile-1595623269.3879838.html
    Some exceptions were thrown in END blocks:
      X::AdHoc: Cannot call method 'nqp-home' on a null object
          in any <main> at /Users/liz/Github/rakudo.moar/install/share/perl6/runtime/perl6.moarvm line 1

and *no* profile file is written.

This seems to have happened just before the 2020.07 release, or shortly after.

**Comment by Altai-man** [comment]
Can confirm a 2020.07 regression, it's a call for a point.

**Comment by AlexDaniel** [comment]
Bisected: (2020-07-16) https://github.com/rakudo/rakudo/commit/1b3780f72569f3a8a74ee4beb69c0a9ca2f2f333
Before and after: https://gist.github.com/8927fa4727f568968d9adfe6625eeabb

Though I think @patrickbkr already knows :)

**Comment by patrickbkr** [comment]
The problem is that the `nqp-home()` method in the `HLL::Compiler` class can not be accessed uniformly. When calling some NQP code the method can be accessed with `nqp::getcomp('nqp').nqp-home()`. When running Raku code one needs to use `nqp::getcomp('Raku').nqp-home()`.

The profiler code in NQP in [src/vm/moar/HLL/Backend.nqp](https://github.com/Raku/nqp/blob/master/src/vm/moar/HLL/Backend.nqp#L702) tries to access the nqp home using the `nqp` compiler object. But the compiler is not registered under that name. So it seems the Compiler object is a poor choice to put that function.

I'm not sure where else we might put it. The code in that function needs to access

- the command line parameters (well, the corresponding `rakudo-home()` function in rakudo does)
- environment variables
- the compile configuration

Where could such a function live?

**Comment by vrurg** [comment]
Can we consider introducing a `'default'` compiler? It's actually not the first time when it doesn't matter what particular compiler `nqp::getcomp` returns as all what's needed is `HLL::Compiler`-declared services.

**Comment by patrickbkr** [comment]
@vrurg I've typed together something along the lines of your idea. It seems to work.

I'm unsure whether the `Compiler` object is the right place to put such generic functionality though. (I don't know of any better place!) Can you review?

**Comment by Altai-man** [comment]
The tests added at https://github.com/rakudo/rakudo/commit/d7aa7e663c07824f26e62689953479208a343822 :

When run from command line using a rakudo with nqp bumped to get the fix:

>rakudo/install/bin/raku -Irakudo/t/packages/ rakudo/t/09-moar/01-profilers.t

It passes properly, but when run as part of test harness using make, the expanded form for a golf:

>/usr/bin/perl -Irakudo/tools/lib -Irakudo/tools/releasable/rakudo/3rdparty/nqp-configure/lib t/harness5 --moar t/09-moar

It fails to create `.html` file and creates `.sql` file instead, so the bug is still around.

**Comment by Altai-man** [comment]
See conversation at https://colabti.org/irclogger/irclogger_log/raku-dev?date=2020-07-27#l322 with some ideas from @timo

**Comment by patrickbkr** [comment]
As far as I can tell, the installed rakudo is fine. It's only the build dir rakudo that still fails to find the profile template.
I'm looking into the build dir rakudo failure now.

**Comment by patrickbkr** [comment]
I think I know what's going on.
The error stems from the fact that in `tools/templates/main-version.in` the `static-nqp-home` is not filled. (That's what timotimo also observed).

The reason for that: `main-version.in` is a backend unspecific template file, but the variable `static_nqp_home` depends on the backend specific nqp configuration. The nqp configuration option is named the same on all backends, so the logic to retrieve it is identical, the value itself as well, but needs to access the backend specific nqp configuration.

The only way I see is to make `main-version.in` backend specific. That's unfortunate as the file is otherwise absolutely backend agnostic.

Then in `Configure.pl` all the logic to set up these configuration variables (in `configure_refine_vars()`) happens before retrieving the nqp configuration (in `gen-nqp` and `configure_active_backends`). So at the point we need the value it's not set up yet.

@vrurg Do you see another way to retrieve a value from the backend nqp configuration that does not force us to turn backend specific?

**Comment by vrurg** [comment]
I didn't closely follow the situation and have no clue as to why exactly `static-nqp-home` affects the ability to find the template. But possibly this is not currently relevant. 

Unfortunately, I also won't have time today to dive deeper into the problem. Hopefully, some overview of the situation would be helpful anyway.

So what we have:

- NQP build is already all backend specific
- _template.html_ is a property of moar backend
- NQP config is not a part of a shared module and this we can only pass values from it into Rakudo at compile time

With all this in mind I don't see why _main-version.nqp_ can't be backend-specific. Basically, we anyway generate it for each backend separately already.

What does raise a flag in this situation is that _template.html_ is being installed in a shared directory without any way to attribute it as a moar-only thing. Would we ever get a profiler for JVM it would have to use a different name limiting our abilities in possibly generalizing the profile-related code in both backends.

There is another solution which appeals to me but requires more thoughtful approach. It was already briefly mentioned before. We can introduce a class at NQP level which would be responsible for system- and backend-dependent data. Similarly to `Compiler` classes, it could be `NQP::SysConfig` serving as a parent for `Rakudo::SysConfig`. This would let us to free ourselves from compile-time only dependency and the need to choose what configuration parameters we actually need to pull from the backend into Rakudo. And, BTW, because an instance of the class would be a singleton anyway, there is no even need to ever instantiate it. Perhaps it'd be sufficient to keep the config data as class locals and provide static methods for accessing it.

Again, it's not been thoroughly thought out. Just a few ideas I had once in a while pulled together.

**Comment by patrickbkr** [comment]
I think I followed a red herring. There is a worse problem.

`nqp/src/HLL/Compiler.nqp` currently contains the `nqp-home()` method. That method needs to access the NQP build configuration, to be precise the `static_nqp_home` variable. The default HLL::Compiler object has an empty `%!config` attribute because the configuration is never initialized. It's not initialized to the NQP configuration nor to the Rakudo configuration. Responsible for the initialization in `NQP::Compiler` is `sub hll-config()`. That sub is not part of the compilation unit containing `HLL::Compiler` (`NQPHLL.nqp`). This hardens my suspicion that `HLL::Compiler` is not the right place to put the nqp-home logic.

So I think I'm back to square one.

Going through the requirements once again. The code in `nqp-home()` needs access to

- environment variables
- the NQP compile configuration (the `static_nqp_home` variable)
- command line parameters (that's not currently needed, but would be wise to make possible. The logic will probably be different in NQP and Rakudo.)

The function or the path it produces should be accessible from anywhere in NQP and Rakudo. Ideally the logic that determines the NQP home is implemented in NQP only and can be used by both, NQP and Rakudo.

Originally the functionality to determine rakudo-home and nqp-home resided in `src/main.nqp` and `src/rakudo-debug.nqp` in Rakudo and set up a HLL symbol using `bindhllsym()`. Now the profiler logic in NQP needs access to the nqp-home variable as well. Following the approach of using a symbol and initializing it as soon as possible in the run means we would have to duplicate the logic in NQP. So we decided to switch over to a method in `HLL::Compiler`.

Maybe the approach of using `bindhllsym` and duplicating the logic in NQP and Rakudo wasn't so bad after all. Downsides:

- Logic to determine the NQP home is duplicated in NQP and Rakudo. (But the logic would be a bit different once we implement overriding using a command line parameter.)
- This approach would mean that every future language built on NQP would need to export a respective symbol as well.
- Coupling of Rakudo and NQP, as NQP depends on the HLL symbol to be exported by Rakudo.

Alternatively we could explore the `NQP::SysConfig` / `Rakudo::SysConfig` option more. One thing that comes to mind that isn't trivial to solve is how Rakudo could override the logic to retrieve a `--nqp-home` command line parameter and make that override have an effect on NQP code using `nqp-home()`.

I'm sorry that I managed to come up with broken solutions to this problem multiple times in a row now. :-(

@vrurg Thanks for providing valuable feedback times and times again. I hope we'll finish this up soon. What's your opinion?

**Comment by vrurg** [comment]
I've got only few minutes today, but the thing about overriding `nqp-home()` method confused me. I don't see a problem to override the method in `Rakudo::SysConfig`. Sure, in NQP you then totally lose access to `NQP::SysConfig` `nqp-home()`, but I see no big deal about it. Say, one would need to split methods into two groups: accessors which can be overriden by children. And all other which must not be overridden. This model is somewhat cumbersome, but that's the limitations of NQP we must live with.

Sorry if I've answered wrong question. Had a hard day and too little time. :)

**Comment by patrickbkr** [comment]
This was an accidental close, probably because GitHub saw a commit already in master a second time via an unrelated PR. I'm reopening.

**Comment by Altai-man** [comment]
We need a PR with tests enabled showing the fix actually works, see https://github.com/rakudo/rakudo/commit/51ea87a3e76324b3c232e26cf5e3e7ea339c17f7
Also, https://colabti.org/irclogger/irclogger_log/raku-dev?date=2020-08-10#l83 ?

**Comment by patrickbkr** [comment]
The breakage nine observed was expected because he tried to build with the nqp PR merged, but not the rakudo PR.
I didn't know about the disabled test. I'll reenable and create another PR. Thanks for the hint!

**Comment by patrickbkr** [comment]
The commit disabling the profile test is now reverted. So I'm only waiting for confirmation from niner whether the build works for him again.

**Comment by niner** [comment]
yep, it's fixed now

**Comment by Altai-man** [comment]
Apparently, the fix does not work on 16.04 ubuntu? https://travis-ci.org/github/nxadm/rakudo-pkg/jobs/723130833

**Comment by nxadm** [comment]
A more static gist:
https://gist.github.com/nxadm/1635b36be8946c9230508aa5314baa47

**Comment by Altai-man** [comment]
UPD: judging by https://travis-ci.org/github/nxadm/rakudo-pkg/builds/723215845 everything is okay except for relocatable builds profile?
This is something to check out before next release.

**Comment by nxadm** [comment]
Hi,

Yes, I discovered this kind of in parallel with you.

Building rakudo from the rakudo repo and making configure build moarvm and nqp results in the failed profile test. Building moarvm, nqp and rakudo forn their respectively repos does not cause the failing test.

**Comment by nxadm** [comment]
I think most people don't build each repo separately, so in that sense it's still a problem on Ubuntu 16.04.

**Comment by Altai-man** [comment]
@nxadm but this is not a blocker for release of "working" packages, right?

**Comment by patrickbkr** [comment]
This error is fixed by PR [#3889](https://github.com/rakudo/rakudo/pull/3889) which hasn't been merged yet.

I'm currently on vacation (until the end of the week) and can't give much support until then. So I'm unsure whether the PR should be merged as soon as possible to have more time before the release, or wait until I'm back and have a little more time at hand should further problems arise.

**Comment by nxadm** [comment]
@Altai-man @patrickbkr I adapted the build script for my relocatable packages, so for my use case I have a perfect workaround. It's still broken for people not building the components separately, but I don't think a dot release is needed as the next release is not that far off.

**Comment by patrickbkr** [comment]
@nxadm I just merged #3889 which should fix the problems reported above with the build time runners not being able to run the profile tests. I'd be grateful for a test if the changes fix the problems on your side.

**Comment by nxadm** [comment]
Thank you,  @patrickbkr. Your stream of fixes really matter for packagers (*and* end users)!
